Multi-step chat with party filtering

emilwareus opened the pull request:
- Added support for dynamic agent approaches in the chat API based on user input and environment variables.
- Introduced integration tests for the chat API to validate multi-step agent functionality and error handling.
- Updated GitHub Actions CI workflow to include Neo4j service setup for testing.
- Modified vitest.config.ts to use the node environment and increased test timeouts for better integration testing.
- Added '.env.test' to .gitignore to prevent tracking of test environment variables.
- Updated package.json and bun.lock to include 'dotenv' for environment variable management.

emilwareus committed:
Enhance chat API and testing setup

- Added support for dynamic agent approaches in the chat API based on user input and environment variables.
- Introduced integration tests for the chat API to validate multi-step agent functionality and error handling.
- Updated GitHub Actions CI workflow to include Neo4j service setup for testing.
- Modified vitest.config.ts to use the node environment and increased test timeouts for better integration testing.
- Added '.env.test' to .gitignore to prevent tracking of test environment variables.
- Updated package.json and bun.lock to include 'dotenv' for environment variable management.

emilwareus commented on src/app/api/chat/__tests__/route.test.ts:
actually testing chat in the CI, with local neo4j deployment and real chat messages to AI

emilwareus commented on src/app/api/chat/route.ts:
```diff
@@ -37,4 +44,99 @@ async function chatHandler(req: NextRequest) {
   }
 }
 
+// Configuration options for different agent approaches
```
Resolve what mode should be used with what configurations. Let's tune this as we go

emilwareus commented on src/lib/gemini-service.ts:
```diff
@@ -15,7 +16,7 @@ if (!API_KEY) {
 
 // Using the model configuration from centralized types
-const LLM_MODEL_KEY: LLMModelKey = 'gemini-1.5-flash-latest';
+const LLM_MODEL_KEY: LLMModelKey = 'gemini-2.5-flash-lite-preview-06-17';
```
New model!

sweoggy replied:
have you compared the bigger gemini-2.5-flash (stable, non-lite) vs this smaller 2.5-flash **_lite_** preview model?

emilwareus replied:
no, not really. Just tested the lite model for now and think it worked fine, but I can play around... this was mainly to solve some stability issues they had on the old lite model for some reason..

sweoggy replied:
These are available as stable models now (without -preview-xx-xx postfix)

emilwareus commented on .github/workflows/ci.yml:
Add neo4j service + env vars with API keys for AI models to enable e2e-style testing of chat

emilwareus commented on .github/workflows/ci.yml:
```diff
@@ -33,24 +59,45 @@ jobs:
+          NEO4J_USERNAME: neo4j
+          NEO4J_PASSWORD: test-password
+          OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
+        run: uv run python main.py --actions graph
```
populate the local database

emilwareus commented on .gitignore:
```diff
@@ -50,3 +50,4 @@ knowledge-base/knowledge_base.egg-info
 knowledge-base/build
 public/kb-documents
 AI_IMPROVEMENTS.md
+.env.test
```
this file is needed for local development now

emilwareus commented on src/lib/multi-step-agent-service.ts:
```diff
@@ -0,0 +1,239 @@
+  } catch (error) {
+    console.error('Error in multi-step agent:', error);
+    
+    // Fallback to basic single query
```
a bit unsure if I like fallbacks like this.. my thought was that in case the LLM dreams up bad filters etc..

emilwareus commented on src/lib/query-generation-service.ts:
```diff
@@ -0,0 +1,256 @@
+  'sverige demokraterna': 'SD',
+};
+
+const QUERY_GENERATION_PROMPT = `Du är en expert på att skapa söktermer för att hitta information om svenska politiska partier och deras ståndpunkter.
```
🇸🇪

emilwareus commented on src/lib/query-generation-service.ts:
```diff
@@ -0,0 +1,256 @@
+    } catch (error) {
+      lastError = error instanceof Error ? error : new Error('Unknown error');
+      
+      // Check if this is a retryable error
+      const isRetryable = lastError.message.includes('overloaded') || 
+                         lastError.message.includes('503') ||
+                         lastError.message.includes('429') ||
+                         lastError.message.includes('timeout');
+      
```
This actually happened time to time with the old 1.5 model..

sweoggy replied:
quite easy to hit the limits if we are on the free tier 😊

emilwareus commented on src/types/model-types.ts:
```diff
@@ -77,6 +77,26 @@ export const LLM_MODELS: Record<string, LLMModelConfig> = {
+    model: 'gemini-2.5-flash-lite-preview-06-17',
+    contextWindow: 1_000_000, // 1M tokens
+    maxOutputTokens: 8192,
+    cost: {
+      inputCostPer1MTokens: {
+        standard: 0.01, // $0.01 per 1M tokens for prompts <= 128k
+        longContext: 0.01  // $0.01 per 1M tokens for prompts > 128k
+      },
+      outputCostPer1MTokens: {
+        standard: 0.04, // $0.04 per 1M tokens for prompts <= 128k
+        longContext: 0.04  // $0.04 per 1M tokens for prompts > 128k
+      },
```
I think we are actually on the free tier

sweoggy replied:
the numbers looks to have too many zeroes?

input cost should be $0.1 per 1M tokens

output $0.4 per 1 M tokens

source https://ai.google.dev/gemini-api/docs/pricing#gemini-2.5-flash-lite

sweoggy commented on src/types/model-types.ts:
```diff
@@ -77,6 +77,26 @@ export const LLM_MODELS: Record<string, LLMModelConfig> = {
+    provider: 'google',
+    model: 'gemini-2.5-flash-lite-preview-06-17',
+    contextWindow: 1_000_000, // 1M tokens
+    maxOutputTokens: 8192,
```
https://ai.google.dev/gemini-api/docs/models#gemini-2.5-flash-lite

```suggestion
    maxOutputTokens: 64000,
```

sweoggy commented on src/types/model-types.ts:
```diff
@@ -77,6 +77,26 @@ export const LLM_MODELS: Record<string, LLMModelConfig> = {
+        longContext: 0.04  // $0.04 per 1M tokens for prompts > 128k
+      },
+      currency: 'USD',
+      lastUpdated: '2024-12-30',
```
I don't know what this is used for, but may as well be set to 2025-06-01

sweoggy commented on src/types/model-types.ts:
```diff
@@ -77,6 +77,26 @@ export const LLM_MODELS: Record<string, LLMModelConfig> = {
+      },
+      currency: 'USD',
+      lastUpdated: '2024-12-30',
+      sourceUrl: 'https://ai.google.dev/gemini-api/docs/pricing'
```
```suggestion
      sourceUrl: 'https://ai.google.dev/gemini-api/docs/pricing#gemini-2.5-flash-lite'
```

emilwareus committed:
Update src/types/model-types.ts

Co-authored-by: Oscar Reimer <oscar.reimer@dmi30.com>

emilwareus committed:
Update src/types/model-types.ts

Co-authored-by: Oscar Reimer <oscar.reimer@dmi30.com>

emilwareus commented on src/lib/gemini-service.ts:
```diff
@@ -15,96 +16,150 @@ if (!API_KEY) {
+const LLM_MODEL_QUERY_KEY: LLMModelKey = 'gemini-2.5-flash-lite-preview-06-17';
 const LLM_CONFIG = LLM_MODELS[LLM_MODEL_KEY];
+const LLM_CONFIG_QUERY = LLM_MODELS[LLM_MODEL_QUERY_KEY];
 const MODEL_NAME = LLM_CONFIG.model;
```
Separated the models, so now lite is used for query generation and flash is used for response @sweoggy

emilwareus reviewed (commented):
Fixed your comments @sweoggy

emilwareus committed:
Refactor Gemini service and multi-step agent to support new query model

- Updated the Gemini service to include a new query model for enhanced context retrieval.
- Modified the multi-step agent service to utilize the new query model for generating search queries, improving flexibility and error handling.
- Adjusted model types to reflect changes in the Gemini model configuration and updated cost parameters.
- Enhanced error handling and tracking for local query generation, ensuring better analytics and performance monitoring.

emilwareus committed:
Update Gemini service model key and remove outdated cost update checks

- Changed the LLM model key in the Gemini service to 'gemini-2.5-flash-lite-preview-06-17' for alignment with the latest model version.
- Removed outdated 'lastUpdated' checks from model cost calculations in tests to streamline validation and ensure accuracy.

sweoggy commented on src/types/model-types.ts:
```diff
@@ -73,8 +73,48 @@ export const LLM_MODELS: Record<string, LLMModelConfig> = {
+      sourceUrl: 'https://ai.google.dev/gemini-api/docs/pricing'
+    }
+  },
+  'gemini-2.5-flash-lite-preview-06-17': {
```
If changing to stable models, don't forget to change this too ☺️

emilwareus committed:
Update Gemini model keys to remove preview version references

- Changed the LLM model key and query key in the Gemini service from 'gemini-2.5-flash-lite-preview-06-17' to 'gemini-2.5-flash-lite' for consistency with the updated model configuration.
- Updated the model configuration in src/types/model-types.ts to reflect the new model key and ensure accurate context and output settings.

sweoggy reviewed (approved):
Solid improvement, thanks!
